core/src/main/java/jenkins/console/ConsoleUrlProvider.java

```diff
@@ -0,0 +1,130 @@
+
+/**
+ * Extension point that allows implementations to redirect build console links to a specified URL.
+ * <p>In order to produce links to console URLs in Jelly templates, use {@link Functions#getConsoleUrl}.
+ * @see Functions#getConsoleUrl
+ * @since TODO
+ */
+public interface ConsoleUrlProvider extends Describable<ConsoleUrlProvider> {
+    @Restricted(NoExternalUse.class)
+    Logger LOGGER = Logger.getLogger(ConsoleUrlProvider.class.getName());
+
+    /**
+     * Get a URL relative to the context path of Jenkins which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the context of Jenkins, or {@code null}
+     * if this implementation does not want to server a special console view for this build.
+     */
+    @CheckForNull String getConsoleUrl(Run<?, ?> run);
+
+    @Override
+    default Descriptor<ConsoleUrlProvider> getDescriptor() {
+        return Jenkins.get().getDescriptorOrDie(getClass());
+    }
+
+    /**
+     * Get a URL relative to the web server root which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * <p>Use {@link Functions#getConsoleUrl} to obtain this link in a Jelly template.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the web server root
+     */
+    static @NonNull String getRedirectUrl(Run<?, ?> run) {
+        ConsoleUrlProviderGlobalConfiguration globalConfig = ConsoleUrlProviderGlobalConfiguration.get();
+        List<ConsoleUrlProvider> providers = globalConfig.getProviders();
+        User currentUser = User.current();
+        if (currentUser != null) { // TODO: Should we add a configuration option to allow admins to suppress this behavior?
+            ConsoleUrlProviderUserProperty userProperty = currentUser.getProperty(ConsoleUrlProviderUserProperty.class);
+            if (userProperty != null) {
+                List<ConsoleUrlProvider> userProviders = userProperty.getProviders();
+                if (userProviders != null) {
+                    providers = userProviders;
+                }
+            }
+        }
+        String url = null;
+        if (providers != null) {
+            for (ConsoleUrlProvider provider : providers) {
+                try {
+                    url = provider.getConsoleUrl(run);
+                    if (url != null) {
+                        break;
+                    }
+                } catch (Exception e) { // Intentionally broad catch clause to guard against broken implementations.
+                    LOGGER.log(Level.WARNING, e, () -> "Error looking up console URL for " + run + " from " + provider.getClass());
+                }
+            }
+        }
+        if (url == null) {
```

Comment:
is this ever hit? won't the default one be hit if there aren't any others?

Reply:
This gets hit if admins never configure anything, or if they do not add the default extension to the global configuration. The default extension is really only useful for individual users in case they do not want to use whatever their admin selects as the default, but it seemed complicated to try to filter out the default extension only in the global case. I can look into that if desired though.

Reply:
Explanatory comment added in [5f051c9](https://github.com/jenkinsci/jenkins/pull/8321/commits/5f051c9b0faf10e39f48122389f38b05ce2b3ef6).